RecoTracker/TkTrackingRegions/plugins/CandidatePointSeededTrackingRegionsProducer.h

```diff
@@ -0,0 +1,451 @@
+  {
+    std::vector<std::unique_ptr<TrackingRegion> > result;
+
+    // pick up the candidate objects of interest
+    edm::Handle< reco::CandidateView > objects;
+    size_t n_objects = 0;
+    size_t n_points = 0;
+
+    if(m_seedingMode == CANDIDATE_SEEDED || m_seedingMode == CANDIDATE_POINT_SEEDED){
+      e.getByToken( token_input, objects );
+      n_objects = objects->size();
+      if (n_objects == 0) return result;
+    }
+
+    if(m_seedingMode == POINT_SEEDED || m_seedingMode == CANDIDATE_POINT_SEEDED){
+      n_points = etaPoints.size();
+      if (n_points == 0) return result;
+    }
+	
+
+    // always need the beam spot (as a fall back strategy for vertex modes)
+    edm::Handle< reco::BeamSpot > bs;
+    e.getByToken( token_beamSpot, bs );
+    if( !bs.isValid() ) return result;
+
+    // this is a default origin for all modes
+    GlobalPoint default_origin( bs->x0(), bs->y0(), bs->z0() );
+
+    // vector of origin & halfLength pairs:
+    std::vector< std::pair< GlobalPoint, float > > origins;
+
+    // fill the origins and halfLengths depending on the mode
+    if (m_mode == BEAM_SPOT_FIXED || m_mode == BEAM_SPOT_SIGMA)
+    {
+      origins.push_back( std::make_pair(
+          default_origin,
+          (m_mode == BEAM_SPOT_FIXED) ? m_zErrorBeamSpot : m_nSigmaZBeamSpot*bs->sigmaZ()
+      ));
+    }
+    else if (m_mode == VERTICES_FIXED || m_mode == VERTICES_SIGMA)
+    {
+      edm::Handle< reco::VertexCollection > vertices;
+      e.getByToken( token_vertex, vertices );
+      int n_vert = 0;
+      for (reco::VertexCollection::const_iterator v = vertices->begin(); v != vertices->end() && n_vert < m_maxNVertices; ++v)
+      {
+        if ( v->isFake() || !v->isValid() ) continue;
+
+        origins.push_back( std::make_pair(
+            GlobalPoint( v->x(), v->y(), v->z() ),
+            (m_mode == VERTICES_FIXED) ? m_zErrorVetex : m_nSigmaZVertex*v->zError()
+        ));
+        ++n_vert;
+      }
+      // no-vertex fall-back case:
+      if (origins.empty())
+      {
+        origins.push_back( std::make_pair(
+            default_origin,
+            (m_nSigmaZBeamSpot > 0.) ? m_nSigmaZBeamSpot*bs->z0Error() : m_zErrorBeamSpot
+        ));
+      }
+    }
+    
+    const MeasurementTrackerEvent *measurementTracker = nullptr;
+    if(!token_measurementTracker.isUninitialized()) {
+      edm::Handle<MeasurementTrackerEvent> hmte;
+      e.getByToken(token_measurementTracker, hmte);
+      measurementTracker = hmte.product();
+    }
+
+    // create tracking regions (maximum MaxNRegions of them) in directions of the
+    // objects of interest (we expect that the collection was sorted in decreasing pt order)
+    int n_regions = 0;
+
+    if(m_seedingMode == CANDIDATE_SEEDED) {
+
+      for(size_t i = 0; i < n_objects && n_regions < m_maxNRegions; ++i ) {
+
+	const reco::Candidate & object = (*objects)[i];
```

Comment:
Would be faster to not go through `edm::Handle` every time, i.e. do e.g. `const auto& objs = *objects;` once and then in the loop ` ... = objs[i];`.